### ISSUE TITLE
MT sliding window algorithm is ineffective

### ISSUE THREAD

**Issue by lhofhansl** [minetest/minetest#7545]
Just for kicks I setup a 100ms delay on my local network. I send observed packet transfer rates.
Turns out that the rate when down to about 5 packets/s (i.e. a RTT of 200ms). So it looks like every single packet is ack'd individually.
There's plenty of CPU on both client and server and plenty of *throughput* on the link. Just latency is high.

That is really bad!
MT should start streaming and delay acks for packets when the RTT is high. The code is clearly there, but it apparently it's not working.

I'll take a look at it during the next few days.

@nerzhul you attempted to replace MT's networking stack, perhaps you have some ideas.

**Comment by lhofhansl** [comment]
@proller do you have some insights from your Freeminer work?

**Comment by sfan5** [comment]
Freeminer just switched to enet AFAIK, nothing that can easily help us fix problems with our network stack.

**Comment by rubenwardy** [comment]
It's worth considering enet again, lots of other people experienced with networking have recommended it to me so I'm not sure that hmmmm was correct

**Comment by nerzhul** [comment]
for me enet has some drawbacks especially in the network loop

**Comment by lhofhansl** [comment]
There's some unnecessary serialization between the packets going on, that's the best explanation why the packet rate goes down significantly as latency increases. It's fairly close 1/RTT packets/s. I feel it's just a bug in the code, but it takes a while to slog through all that code.

**Comment by lhofhansl** [comment]
Just for my own reference. When I turn on logging:
```
2018-07-12 11:45:36: INFO[ConnectionSend]: con(26/2)     channel: 2, peer quota:1024
2018-07-12 11:45:36: INFO[ConnectionSend]:                      reliables on wire: 4, waiting for ack for 105
2018-07-12 11:45:36: INFO[ConnectionSend]:                      incoming_reliables: 0, next reliable packet: 419, next queued: 0
2018-07-12 11:45:36: INFO[ConnectionSend]:                      reliables queued : 2
2018-07-12 11:45:36: INFO[ConnectionSend]:                      queued commands  : 237
```
Note that only 4 reliables are on the wire. This varies between 3 and 8.

**Comment by lhofhansl** [comment]
Hah... Found the problem. When we disable the legacy peer setting we forget to set it on the local side as well.
This fixes it for me, but I'm not sure it's quite correct. Could also be the client misses sending a crucial packet.

```
diff --git a/src/network/connectionthreads.cpp b/src/network/connectionthreads.cpp
index c4e51211..97af52c3 100644
--- a/src/network/connectionthreads.cpp
+++ b/src/network/connectionthreads.cpp
@@ -1215,6 +1215,7 @@ SharedBuffer<u8> ConnectionReceiveThread::handlePacketType_Control(Channel *chan
                writeU8(&reply[1], CONTROLTYPE_ENABLE_BIG_SEND_WINDOW);
                cmd.disableLegacy(PEER_ID_SERVER, reply);
                m_connection->putCommand(cmd);
+               dynamic_cast<UDPPeer *>(peer)->setNonLegacyPeer();
 
                throw ProcessedSilentlyException("Got a SET_PEER_ID");
        } else if (controltype == CONTROLTYPE_PING) {
```

**Comment by lhofhansl** [comment]
Fixed with #7548 and #7549
